File: 9. Merge K sorted linked lists using min-heap.py
"""
Task: Merge K sorted linked lists using a min-heap
Time Complexity: O(N log k)
Space Complexity: O(k + N)
"""

import heapq


class Node:
    def __init__(self, val):
        self.val = val
        self.next = None


# Build three example sorted lists
l1 = Node(1)
l1.next = Node(4)
l1.next.next = Node(7)

l2 = Node(2)
l2.next = Node(5)
l2.next.next = Node(8)

l3 = Node(3)
l3.next = Node(6)
l3.next.next = Node(9)

lists = [l1, l2, l3]

pq = []
counter = 0
for head in lists:
    if head:
        heapq.heappush(pq, (head.val, counter, head))
        counter += 1

dummy = Node(0)
tail = dummy
while pq:
    val, _, node = heapq.heappop(pq)
    tail.next = Node(node.val)
    tail = tail.next
    if node.next:
        heapq.heappush(pq, (node.next.val, counter, node.next))
        counter += 1

res = []
it = dummy.next
while it:
    res.append(it.val)
    it = it.next

print('Merged list:', res)

File: 9. Merge K sorted linked lists using min-heap.cpp
/*
 Task: Merge K sorted linked lists using a min-heap
*/

#include <bits/stdc++.h>
using namespace std;

struct Node
{
    int val;
    Node* next;
    Node(int v) : val(v), next(nullptr) {}
};
struct Cmp
{
    bool operator()(Node* a, Node* b) const { return a->val > b->val; }
};

int main()
{
    // build three small lists
    Node* a = new Node(1);
    a->next = new Node(4);
    a->next->next = new Node(7);
    Node* b = new Node(2);
    b->next = new Node(5);
    b->next->next = new Node(8);
    Node* c = new Node(3);
    c->next = new Node(6);
    c->next->next = new Node(9);

    vector<Node*> lists = {a, b, c};
    priority_queue<Node*, vector<Node*>, Cmp> pq;
    for (Node* L : lists)
        if (L)
            pq.push(L);

    Node dummy(0);
    Node* tail = &dummy;
    while (!pq.empty())
    {
        Node* t = pq.top();
        pq.pop();
        tail->next = new Node(t->val);
        tail = tail->next;
        if (t->next)
            pq.push(t->next);
    }

    cout << "Merged list: ";
    Node* it = dummy.next;
    while (it)
    {
        cout << it->val << " ";
        it = it->next;
    }
    cout << '\n';
    return 0;
}

// Time Complexity: O(N log k)
// Space Complexity: O(k + N)
